File: reimplementation/adl2/crates/adl-certify/src/bin/recheck.rs
//! `smash2-recheck` — standalone replay of `--combine` certificate bundles.
//!
//! Re-checks each bundle with the trusted exact-rational kernel only: no
//! solver, no search, no smash2 analysis run. A bundle that replays proves
//! its listed formulas are (real-)unsatisfiable together, and re-derives every
//! reconciliation fact it uses from that fact's own embedded premises — the
//! bundle's `note` field states what that does and does not cover, and this
//! tool repeats the part people most often over-read (see [`CAVEAT`]).
//!
//! Bundles written under the superseded schema `smash2-combine/1` are
//! **rejected**, not half-checked: they carry no derivation chain, so
//! "OK" would mean something weaker than it does for a `/2` bundle. Re-emit
//! them with a current smash2.
//!
//! Usage: `smash2-recheck BUNDLE.json... | DIR...`  (a directory contributes
//! its `*.json` files, sorted).
//!
//! # Exit codes
//!
//! | code | meaning |
//! |------|---------|
//! | 0 | at least one bundle was checked and **every** one replayed |
//! | 1 | bundles were checked and at least one failed to replay |
//! | 2 | nothing could be checked: bad usage, an unreadable path, or a
//!       directory that contains no `*.json` bundles |
//!
//! Exit 2 for "no bundles found" is deliberate and must stay non-zero: a
//! release script that runs `smash2-recheck dist/bundles/` to gate a publish
//! has to fail when the bundle directory is empty, not silently pass having
//! verified nothing. Vacuous success is the one outcome a verification tool
//! must never produce.

use adl_certify::CombineBundle;
use adl_certify::bundle::{BUNDLE_SCHEMA, SUPERSEDED_SCHEMAS, supersession_note};
use serde::Deserialize;
use std::path::PathBuf;
use std::process::ExitCode;

/// What a passing replay does NOT establish. Printed with the summary so an
/// `OK` line is never read as an endorsement of the bundle's prose.
const CAVEAT: &str = "replay proves formula-level unsatisfiability and the derivation \
     chain of every reconciliation fact used; quantity labels, assert sources, region \
     names, producer and input identity are unauthenticated description and are NOT \
     checked - see each bundle's `note`";

/// Just enough of a bundle to report a schema mismatch precisely: a `/1`
/// document does not deserialize into the current struct at all, so the schema
/// has to be read before the full parse.
#[derive(Deserialize)]
struct SchemaProbe {
    schema: String,
}

fn collect_inputs(args: &[String]) -> Result<Vec<PathBuf>, String> {
    let mut out = Vec::new();
    for a in args {
        let p = PathBuf::from(a);
        if p.is_dir() {
            let mut found: Vec<PathBuf> = std::fs::read_dir(&p)
                .map_err(|e| format!("cannot read directory {}: {e}", p.display()))?
                .filter_map(Result::ok)
                .map(|e| e.path())
                .filter(|q| q.extension().is_some_and(|x| x == "json"))
                .collect();
            found.sort();
            if found.is_empty() {
                let total = std::fs::read_dir(&p).map(Iterator::count).unwrap_or(0);
                return Err(format!(
                    "no *.json certificate bundles in directory {} ({}); \
                     nothing was verified, so this exits 2 rather than reporting success. \
                     Bundles are produced by `smash2 verify --combine {}`",
                    p.display(),
                    match total {
                        0 => "the directory is empty".to_owned(),
                        n => format!("{n} entr{} present, none ending in .json", {
                            if n == 1 { "y" } else { "ies" }
                        }),
                    },
                    p.display(),
                ));
            }
            out.extend(found);
        } else {
            out.push(p);
        }
    }
    Ok(out)
}

fn check(path: &PathBuf) -> Result<String, String> {
    let text = std::fs::read_to_string(path).map_err(|e| format!("read error: {e}"))?;
    let probe: SchemaProbe =
        serde_json::from_str(&text).map_err(|e| format!("parse error: {e}"))?;
    if SUPERSEDED_SCHEMAS.contains(&probe.schema.as_str()) {
        return Err(supersession_note(&probe.schema));
    }
    if probe.schema != BUNDLE_SCHEMA {
        // The schema string is attacker-controlled and unbounded; quote only a
        // recognizable prefix rather than echoing megabytes back to the reader.
        let shown: String = probe.schema.chars().take(64).collect();
        let ellipsis = if shown.len() < probe.schema.len() {
            "..."
        } else {
            ""
        };
        return Err(format!(
            "unknown schema {shown:?}{ellipsis} (expected {BUNDLE_SCHEMA:?})"
        ));
    }
    let bundle: CombineBundle =
        serde_json::from_str(&text).map_err(|e| format!("parse error: {e}"))?;
    if !bundle.replay() {
        return Err("certificate does not refute the listed formulas, or a \
                    reconciliation fact is used without a replayable derivation"
            .to_owned());
    }
    let chain = match bundle.derived_facts.len() {
        0 => String::new(),
        n => format!(", {n} derived fact(s) re-derived"),
    };
    Ok(format!("{} vs {}{chain}", bundle.region_a, bundle.region_b))
}

/// Restore the default SIGPIPE disposition, so `smash2-recheck DIR/ | head`
/// terminates on the closed pipe instead of panicking out of `println!`.
/// See `adl-cli`'s `restore_sigpipe` for the full rationale and tradeoff.
#[cfg(unix)]
fn restore_sigpipe() {
    // SAFETY: runs once at the top of `main`, before any thread exists.
    unsafe {
        libc::signal(libc::SIGPIPE, libc::SIG_DFL);
    }
}

#[cfg(not(unix))]
fn restore_sigpipe() {}

fn main() -> ExitCode {
    restore_sigpipe();
    let args: Vec<String> = std::env::args().skip(1).collect();
    if args.is_empty() || args.iter().any(|a| a == "--help" || a == "-h") {
        eprintln!("usage: smash2-recheck BUNDLE.json... | DIR...");
        eprintln!("re-checks smash2 `verify --combine` certificate bundles ({BUNDLE_SCHEMA})");
        eprintln!("with the trusted exact-rational kernel; no solver required.");
        eprintln!();
        eprintln!("exit codes:");
        eprintln!("  0  every bundle checked replayed successfully (at least one was checked)");
        eprintln!("  1  a bundle failed to replay");
        eprintln!("  2  nothing was checked: bad usage, unreadable path, or a directory");
        eprintln!("     with no *.json bundles (fail-closed, so an empty bundle directory");
        eprintln!("     never gates a release as a vacuous success)");
        return ExitCode::from(2);
    }

    let files = match collect_inputs(&args) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("smash2-recheck: {e}");
            return ExitCode::from(2);
        }
    };

    let mut checked = 0usize;
    let mut failed = 0usize;
    for f in &files {
        checked += 1;
        match check(f) {
            Ok(what) => println!("OK   {} ({what})", f.display()),
            Err(why) => {
                failed += 1;
                println!("FAIL {} — {why}", f.display());
            }
        }
    }

    println!("{checked} bundle(s) checked, {failed} failed — {CAVEAT}");
    if failed == 0 && checked > 0 {
        ExitCode::SUCCESS
    } else {
        ExitCode::FAILURE
    }
}

File: reimplementation/smash_cpp2/libs/cli/src/recheck.cpp
#include "adl2/certify/bundle.hpp"

#include <algorithm>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

const char* CAVEAT =
    "replay proves formula-level unsatisfiability and the derivation "
    "chain of every reconciliation fact used; quantity labels, assert sources, region "
    "names, producer and input identity are unauthenticated description and are NOT "
    "checked - see each bundle's `note`";

std::string read_file(const std::string& path) {
  std::ifstream in(path);
  if (!in) return {};
  std::ostringstream ss;
  ss << in.rdbuf();
  return ss.str();
}

std::string schema_of(const std::string& text) {
  auto b = adl2::certify::CombineBundle::from_json(text);
  if (b) return b->schema;
  const std::string key = "\"schema\"";
  auto pos = text.find(key);
  if (pos == std::string::npos) return {};
  pos = text.find('"', pos + key.size());
  if (pos == std::string::npos) return {};
  auto end = text.find('"', pos + 1);
  if (end == std::string::npos) return {};
  return text.substr(pos + 1, end - pos - 1);
}

int collect_inputs(const std::vector<std::string>& args, std::vector<std::string>& out,
                   std::string& err) {
  for (const auto& a : args) {
    std::error_code ec;
    std::filesystem::path p(a);
    if (std::filesystem::is_directory(p, ec)) {
      std::vector<std::string> found;
      for (const auto& ent : std::filesystem::directory_iterator(p, ec)) {
        if (ec) {
          err = "cannot read directory " + a + ": " + ec.message();
          return 2;
        }
        if (ent.path().extension() == ".json") found.push_back(ent.path().string());
      }
      std::sort(found.begin(), found.end());
      if (found.empty()) {
        std::size_t total = 0;
        std::error_code ec2;
        for (const auto& ent : std::filesystem::directory_iterator(p, ec2)) {
          (void)ent;
          ++total;
        }
        err = "no *.json certificate bundles in directory " + a + " (" +
              (total == 0 ? std::string("the directory is empty")
                          : std::to_string(total) + " entries present, none ending in .json") +
              "); nothing was verified, so this exits 2 rather than reporting success. "
              "Bundles are produced by `smash_cpp2 verify --combine " +
              a + "`";
        return 2;
      }
      out.insert(out.end(), found.begin(), found.end());
    } else {
      out.push_back(a);
    }
  }
  return 0;
}

int check(const std::string& path, std::string& what, std::string& why) {
  std::ifstream probe(path);
  if (!probe) {
    why = "read error: cannot open " + path;
    return 1;
  }
  probe.close();
  std::string text = read_file(path);
  std::string schema = schema_of(text);
  if (schema == adl2::certify::SUPERSEDED_SCHEMA_V1) {
    why = adl2::certify::supersession_note(schema);
    return 1;
  }
  auto bundle = adl2::certify::CombineBundle::from_json(text);
  if (!bundle) {
    if (!schema.empty() && schema != adl2::certify::BUNDLE_SCHEMA) {
      std::string shown = schema.substr(0, 64);
      std::string ellipsis = schema.size() > 64 ? "..." : "";
      why = "unknown schema \"" + shown + ellipsis + "\" (expected \"" +
            std::string(adl2::certify::BUNDLE_SCHEMA) + "\")";
    } else {
      why = "parse error: not a smash2-combine/2 document";
    }
    return 1;
  }
  if (bundle->schema != adl2::certify::BUNDLE_SCHEMA) {
    std::string shown = bundle->schema.substr(0, 64);
    std::string ellipsis = bundle->schema.size() > 64 ? "..." : "";
    why = "unknown schema \"" + shown + ellipsis + "\" (expected \"" +
          std::string(adl2::certify::BUNDLE_SCHEMA) + "\")";
    return 1;
  }
  if (!bundle->replay()) {
    why = "certificate does not refute the listed formulas, or a "
          "reconciliation fact is used without a replayable derivation";
    return 1;
  }
  what = bundle->region_a + " vs " + bundle->region_b;
  if (!bundle->derived_facts.empty()) {
    what += ", " + std::to_string(bundle->derived_facts.size()) + " derived fact(s) re-derived";
  }
  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  std::vector<std::string> args;
  for (int i = 1; i < argc; ++i) args.emplace_back(argv[i]);
  if (args.empty() || std::find(args.begin(), args.end(), "--help") != args.end() ||
      std::find(args.begin(), args.end(), "-h") != args.end()) {
    std::cerr << "usage: smash_cpp2-recheck BUNDLE.json... | DIR...\n";
    std::cerr << "re-checks smash_cpp2 `verify --combine` certificate bundles ("
              << adl2::certify::BUNDLE_SCHEMA << ")\n";
    std::cerr << "with the trusted exact-rational kernel; no solver required.\n\n";
    std::cerr << "exit codes:\n";
    std::cerr << "  0  every bundle checked replayed successfully (at least one was checked)\n";
    std::cerr << "  1  a bundle failed to replay\n";
    std::cerr << "  2  nothing was checked: bad usage, unreadable path, or a directory\n";
    std::cerr << "     with no *.json bundles (fail-closed, so an empty bundle directory\n";
    std::cerr << "     never gates a release as a vacuous success)\n";
    return 2;
  }

  std::vector<std::string> files;
  std::string err;
  if (collect_inputs(args, files, err) != 0) {
    std::cerr << "smash_cpp2-recheck: " << err << "\n";
    return 2;
  }

  std::size_t checked = 0;
  std::size_t failed = 0;
  for (const auto& f : files) {
    ++checked;
    std::string what;
    std::string why;
    if (check(f, what, why) == 0) {
      std::cout << "OK   " << f << " (" << what << ")\n";
    } else {
      ++failed;
      std::cout << "FAIL " << f << " — " << why << "\n";
    }
  }

  std::cout << checked << " bundle(s) checked, " << failed << " failed — " << CAVEAT << "\n";
  if (failed == 0 && checked > 0) return 0;
  return 1;
}
